compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -109,45 +126,159 @@ impl InternalAudioMixer {
-                let time_to_remove_from_end = input_samples.end_pts.saturating_sub(end_pts);
-                let samples_after_frame =
-                    (time_to_remove_from_end.as_secs_f64() * sample_rate as f64).ceil() as usize;
-                let end_index = input_samples.len() - samples_after_frame;
-
-                samples_in_frame.extend(input_samples.samples[start_index..end_index].iter());
-
-                input_samples.end_pts
-            });
-        // Appending samples missing
-        let missing_samples = Self::samples_in_frame(start_pts, end_pts, sample_rate);
-        Self::push_missing_samples(&mut samples_in_frame, missing_samples);
+        // Real numerical errors are a lot smaller, but taking max error as 1% of a sample duration
+        // seems to be safe enough.
+        let max_error = Duration::from_secs_f64(0.01 / sample_rate as f64);
+
+        // Output and input samples have the same sample rate, but they are not synced with each
+        // other. We need to calculate an offset between input and output samples. This value
+        // should be constant for a specific input, but there is no harm with calculating it every
+        // time.
+        let sample_offset = samples
+            .first()
+            .map(|batch| {
+                let duration_secs = start_pts.as_secs_f64() - batch.start_pts.as_secs_f64();
+                let sample_duration_secs = 1.0 / sample_rate as f64;
+                Duration::from_secs_f64(duration_secs.rem_euclid(sample_duration_secs))
+            })
+            .unwrap_or(Duration::ZERO);
+
+        let time_to_sample_count = |duration: Duration| {
+            let sample_count = duration.as_secs_f64() * sample_rate as f64;
+            if (sample_count - sample_count.round()).abs() < max_error.as_secs_f64() {
+                sample_count.round() as usize
+            } else {
+                sample_count.floor() as usize
+            }
+        };
+
+        let last_batch_end_pts = samples.last().map(|sample| sample.end_pts + sample_offset);
+        let samples_iter = samples.into_iter().map(|mut sample| {
+            sample.start_pts += sample_offset;
+            sample.end_pts += sample_offset;
+            sample
+        });
+
+        for (batch_index, input_samples) in samples_iter.enumerate() {
+            let sample_count = samples_in_frame.len();
+            let expected_next_sample_start_pts =
+                start_pts + Duration::from_secs_f64(sample_count as f64 / sample_rate as f64);
+
+            // potentially fill missing spots
+            if expected_next_sample_start_pts + max_error < input_samples.start_pts {
+                let missing_time = input_samples
+                    .start_pts
+                    .saturating_sub(expected_next_sample_start_pts);
+                let missing_samples_count = time_to_sample_count(missing_time);
+                if missing_samples_count < 1 {
+                    warn!(
```

Comment:
if distance is larger than max_error but less than one sample then most likely this is a bug